client/src/cbltest/api/couchbaseserver.py

```
@@ -49,8 +102,21 @@ def __init__(self, url: str, username: str, password: str):
             opts = ClusterOptions(auth)
             self.__username = username
             self.__password = password
-            self.__cluster = Cluster(url, opts)
-            self.__cluster.wait_until_ready(timedelta(seconds=10))
+
+            try:
+                self.__cluster = Cluster(url, opts)
+                self.__cluster.wait_until_ready(timedelta(seconds=10))
+            except Exception:
+                self.__cluster = None
```

If you do this, suddenly you need null checks everywhere.  Why not just let the exception propagate if this fails?

---

This returns the status of the cluster, to return the __cluster = None here is one way of checking and assigning each of the nodes their status. See the issue is I have to add such checks in case a test fails midway, and then the next test has to run which again will use all the nodes (or even some of them), in that case the next test first checks for the health of all the nodes, and in case there is a node unresponsive of not part of the cluster due to a midway failure, it then runs a full cluster setup again before running the actual test.

---

The other solution that looks graceful is adding the node health checks to be in cluster,  at the fixture level in conftest.py but that is wasteful since it would then run before every test and its better to manually call it only in the multi CBS node tests like in the test file at line : 102 `cbs_one.ensure_cluster_healthy(cblpytest.couchbase_servers)`

Similarly at 252, if first test fails after node rebalance out , the second one runs an **all node health check first.**

Trust me I tried other ways via chatbot too but they involved more clutter and this seems the least cluttering.
ensure_cluster_healthy -> This just makes rest calls to the cluster and then each node to see if its status is one of `Active` and `inactiveFailed`, and call `recover` or `addNode` respectively

---

Makes sense, actually yes I didn't think of it like that, and since all the mypy checks were passing (which make sure a possibly `NoneType` to be checked), I didn't bother.
Even in the tests it behaved fine, which I ran multiple times, but you're right to have a **FAIL FAST** to be there and halt at a failed initialization of cluster.
Done ✅